bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/service/BusinessPartnerService.kt

```diff
@@ -49,110 +50,137 @@ class BusinessPartnerService(
+            .ifEmpty { listOf(null) }
 
-    fun getBusinessPartnersInput(pageRequest: PageRequest, externalIds: Collection<String>?): PageDto<BusinessPartnerInputDto> {
-        val stage = StageType.Input
-        val page = when {
-            externalIds.isNullOrEmpty() -> businessPartnerRepository.findByStage(stage, pageRequest)
-            else -> businessPartnerRepository.findByStageAndExternalIdIn(stage, externalIds, pageRequest)
-        }
-        return page.toPageDto(businessPartnerMappings::toBusinessPartnerInputDto)
+        businessPartnerTypes.forEach { type -> changelogRepository.save(ChangelogEntry(partner.externalId, type, changelogType, partner.stage)) }
     }
 
-    fun getBusinessPartnersOutput(pageRequest: PageRequest, externalIds: Collection<String>?): PageDto<BusinessPartnerOutputDto> {
-        val stage = StageType.Output
-        val page = when {
-            externalIds.isNullOrEmpty() -> businessPartnerRepository.findByStage(stage, pageRequest)
-            else -> businessPartnerRepository.findByStageAndExternalIdIn(stage, externalIds, pageRequest)
+    private fun assertInputStageExists(externalIds: Collection<String>) {
+        val existingExternalIds = businessPartnerRepository.findByStageAndExternalIdIn(StageType.Input, externalIds)
+            .map { it.externalId }
+            .toSet()
+
+        if (externalIds.minus(existingExternalIds).isNotEmpty()) {
+            val missing = externalIds.minus(existingExternalIds)
+            throw BpdmBusinessPartnerStageNotExistsException(missing.first(), StageType.Input)
         }
-        return page.toPageDto(businessPartnerMappings::toBusinessPartnerOutputDto)
     }
 
-    private fun getExistingEntityByExternalId(
-        dtos: Collection<IBaseBusinessPartnerGateDto>,
-        stage: StageType
-    ): Map<String, BusinessPartner> {
-        val externalIds = dtos.map { it.externalId }.toSet()
-        return businessPartnerRepository.findByStageAndExternalIdIn(stage, externalIds)
+    private fun replaceWithExistingAndPartition(entityCandidates: List<BusinessPartner>, stage: StageType):
+            UpsertPartition {
+        val existingPartnersByExternalId = businessPartnerRepository.findByStageAndExternalIdIn(stage, entityCandidates.map { it.externalId })
             .associateBy { it.externalId }
-    }
 
-    private fun initSharingState(dto: IBaseBusinessPartnerGateDto) {
-        // TODO make businessPartnerType optional
-        sharingStateService.upsertSharingState(SharingStateDto(BusinessPartnerType.ADDRESS, dto.externalId))
+        val newEntities = mutableListOf<BusinessPartner>()
+        val existingEntities = mutableListOf<BusinessPartner>()
+
+        entityCandidates.forEach { candidate ->
+            val existingEntity = existingPartnersByExternalId[candidate.externalId]
+            if (existingEntity != null)
+                existingEntities.add(copyValues(candidate, existingEntity))
```

Comment:
Could we do the saveChangelog here? Then it is maybe not neccessary to split in new and existing entities.

Reply:
I reworked this and now create a new internal data class for entity resolution results and this is then used for creating the changelogs